schema/function_test.go

```diff
@@ -0,0 +1,664 @@
+	assert.NoError(t, err)
+	assert.Equals[schema.Type](t, typeOutput, schema.NewListSchema(
+		schema.NewStringSchema(nil, nil, nil),
+		nil, nil,
+	))
+	assert.Equals(t, errOut, true)
+	// Validate that the function call works as intended
+	result, err := simpleFunc.Call([]any{"a"})
+	assert.NoError(t, err)
+	assert.Equals(t, result.([]string), []string{"a", "a"})
+
+	// Validate type
+	typeOutput, errOut, err = simpleFunc.Output([]schema.Type{schema.NewIntSchema(nil, nil, nil)})
+	assert.NoError(t, err)
+	assert.Equals[schema.Type](t, typeOutput, schema.NewListSchema(
+		schema.NewIntSchema(nil, nil, nil),
+		nil, nil,
+	))
+	assert.Equals(t, errOut, true)
+	// Validate that the function call works as intended
+	result, err = simpleFunc.Call([]any{1})
+	assert.NoError(t, err)
+	assert.Equals(t, result.([]int), []int{1, 1})
+
+	// Validate type
+	typeOutput, errOut, err = simpleFunc.Output([]schema.Type{
+		schema.NewListSchema(schema.NewIntSchema(nil, nil, nil), nil, nil),
+	})
+	assert.NoError(t, err)
+	assert.Equals[schema.Type](t, typeOutput, schema.NewListSchema(
+		schema.NewListSchema(
+			schema.NewIntSchema(nil, nil, nil), nil, nil),
+		nil, nil),
+	)
+	assert.Equals(t, errOut, true)
+	// Validate that the function call works as intended
+	result, err = simpleFunc.Call([]any{[]int{1}})
+	assert.NoError(t, err)
+	assert.Equals(t, result.([][]int), [][]int{{1}, {1}})
+}
+
+func TestFunctionToStringOneParamVoid(t *testing.T) {
+	oneParamVoidFunction, err := schema.NewCallableFunction(
+		"a",
+		[]schema.Type{schema.NewStringSchema(nil, nil, nil)},
+		nil,
+		false,
+		nil,
+		func(a string) {},
+	)
+	assert.NoError(t, err)
+	funcStr := oneParamVoidFunction.String()
+	assert.Equals(t, funcStr, "a(string) void")
+}
+
+func TestFunctionToStringToFunctionTwoParam(t *testing.T) {
+	callableFunc, err := schema.NewCallableFunction(
+		"b",
+		[]schema.Type{
+			schema.NewStringSchema(nil, nil, nil),
+			schema.NewIntSchema(nil, nil, nil),
+		},
+		schema.NewIntSchema(nil, nil, nil),
+		false,
+		nil,
+		func(a string, b int64) int64 { return 0 },
+	)
+	assert.NoError(t, err)
+	funcStr := callableFunc.String()
+	assert.Equals(t, funcStr, "b(string, integer) integer")
+
+	// Now the version in FunctionSchema instead of CallableFunction
+	funcSchema, err := callableFunc.ToFunctionSchema()
+	assert.NoError(t, err)
+
+	funcStr = funcSchema.String()
+	assert.Equals(t, funcStr, "b(string, integer) integer")
+}
+
+func TestDynamicFunctionToStringToFunction(t *testing.T) {
+	oneParamVoidFunction, err := schema.NewDynamicCallableFunction(
+		"c",
+		[]schema.Type{schema.NewAnySchema(), schema.NewIntSchema(nil, nil, nil)},
+		nil,
+		func(a any, b int64) (any, error) {
+			return b, nil
+		},
+		func(inputType []schema.Type) (schema.Type, error) {
+			return schema.NewIntSchema(nil, nil, nil), nil
+		},
+	)
+	assert.NoError(t, err)
+	funcStr := oneParamVoidFunction.String()
+	assert.Equals(t, funcStr, "c(any, integer) (dynamic, error)")
```

Comment:
Do dynamic functions always/have-to return an error value?  (If not, should you have a test of one that doesn't?)

Reply:
They should. It's enforced by NewDynamicCallableFunction.